dev/com.ibm.ws.messaging.open_jms20_fat/fat/src/com/ibm/ws/messaging/JMS20/fat/SharedSubscription/SharedSubscriptionWithMsgSelTest_129623.java

```diff
@@ -85,27 +79,33 @@ private static String getLocalAddress() {
     }
 
     private static boolean runInServlet(String test) throws IOException {
-        return TestUtils.runInServlet( clientHostName, clientPort, subscriptionContextRoot, test, getLocalAddress() );
+        return TestUtils.runInServlet(clientHostName, clientPort, subscriptionContextRoot, test, getLocalAddress());
         // throws IOException
     }
 
-    //
-
     @BeforeClass
     public static void testConfigFileChange() throws Exception {
         engineServer.copyFileToLibertyInstallRoot(
-            "lib/features",
-            "features/testjmsinternals-1.0.mf");
+                                                  "lib/features",
+                                                  "features/testjmsinternals-1.0.mf");
         engineServer.setServerConfigurationFile("SharedSubscriptionEngine.xml");
 
         clientServer.copyFileToLibertyInstallRoot(
-            "lib/features",
-            "features/testjmsinternals-1.0.mf");
+                                                  "lib/features",
+                                                  "features/testjmsinternals-1.0.mf");
         TestUtils.addDropinsWebApp(clientServer, subscriptionAppName, subscriptionPackages);
-        clientServer.setServerConfigurationFile("SharedSubscriptionDurClient.xml");
+        final String clientXml = "SharedSubscriptionDurClient.xml";
+        if (JakartaEE9Action.isActive()) {
+            Path clientXmlFile = Paths.get(
+                                           "lib/LibertyFATTestFiles",
+                                           clientXml);
+            JakartaEE9Action.transformApp(clientXmlFile);
+            Log.info(c, "setUp", "Transformed server " + clientXmlFile);
+        }
+        clientServer.setServerConfigurationFile(clientXml);
```

Comment:
Doesn't the usual server transformation steps handle this file?  Or is it not transformed because it is not under the 'servers' folder?  Also, there are some helper methods, either on LibertyServer or on ShrinkWrapHelper which may do something similar.

Reply:
Simplicity does not invoke the transformer on server.xml files, at least not yet.  The most direct, simple solution is to invoke the transformer directly on the server configuration file and include the file name and transformer properties in the transformer property set.  That's what I've done here.  Simplicity can be modified to transform server and client configuration files, and there a couple ways this can be done.  Simplicity could scan the publish/ and publish/server/ directories for xml files and transform all files that match file names declared in a repeat action builder method, where the transformer is called with arg1=properties file containing the file name(s) and arg2=properties file containig match-replace values for all config metatypes.  Alternatively, Simplicity could directly replace matching strings in the ServerConfiguration instance that is instantiated in order to replace features.  When the instance is streamed to file all necessary transformation has been done -- i.e., there's no need to invoke the transformer , which is expensive.

Reply:
Thanks for the heads up!  I have a PR that does option 1 (calls the transformer on server.xml) here: https://github.com/OpenLiberty/open-liberty/pull/14397  but as you mentioned this is an expensive operation to do for each server.xml.  However, it does offer us more control in what we are replacing since not ever instance of `javax` should be replaced (i.e. javax.sql.XADataSource)

Reply:
Yes, that the idea.  Augment the ee9 repeat action to replace strings in config elements other than <FeatureManager> according to, say a new builder method, before the ServerConfiguration is streamed to file.  Any solution will have full control over what is transformed.  Even if the repeat action is modified to perform separate operations to replace features and other config, the two operations are still much less costly than invoking the transformer.  Something to consider.  What I like about using the transformer is the consistency of using it for both config metatype metadata and the server/client configuration metadata.  Thanks!